File: Easy/Hashing/205_Isomorphic_Strings.cpp
/*
Problem: Isomorphic Strings

Approach:
- Hashing using two unordered maps
- Maintain bidirectional character mapping between both strings

Time Complexity: O(n)
Space Complexity: O(n)
*/

#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        if (s.length() != t.length()) return false;

        unordered_map<char, char> mapST;
        unordered_map<char, char> mapTS;

        for (int i = 0; i < s.length(); i++) {
            char cs = s[i];
            char ct = t[i];

            // s -> t mapping check
            if (mapST.count(cs)) {
                if (mapST[cs] != ct)
                    return false;
            } else {
                mapST[cs] = ct;
            }
            // t -> s mapping check
            if (mapTS.count(ct)) {
                if (mapTS[ct] != cs)
                    return false;
            } else {
                mapTS[ct] = cs;
            }
        }
        return true;
    }
};

File: Easy/Hashing/205_Isomorphic_Strings.py
'''
Problem: Isomorphic Strings

Approach:
- Hashing using two unordered maps
- Maintain bidirectional character mapping between both strings

Time Complexity: O(n)
Space Complexity: O(n)
'''

class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        s_to_t = {}
        t_to_s = {}

        for i in range(len(s)):
            c1 = s[i]
            c2 = t[i]

            if c1 in s_to_t:
                if s_to_t[c1] != c2:
                    return False
            else:
                s_to_t[c1] = c2

            if c2 in t_to_s:
                if t_to_s[c2] != c1:
                    return False
            else:
                t_to_s[c2] = c1

        return True
